core/src/main/java/com/sequenceiq/cloudbreak/conclusion/step/SaltCheckerConclusionStep.java

```diff
@@ -0,0 +1,118 @@
+import com.cloudera.thunderhead.telemetry.nodestatus.NodeStatusProto.SaltMinionsHealth;
+import com.cloudera.thunderhead.telemetry.nodestatus.NodeStatusProto.ServiceStatus;
+import com.cloudera.thunderhead.telemetry.nodestatus.NodeStatusProto.StatusDetails;
+import com.sequenceiq.cloudbreak.client.RPCResponse;
+import com.sequenceiq.cloudbreak.common.exception.CloudbreakServiceException;
+import com.sequenceiq.cloudbreak.domain.stack.Stack;
+import com.sequenceiq.cloudbreak.node.status.NodeStatusService;
+import com.sequenceiq.cloudbreak.orchestrator.model.Node;
+import com.sequenceiq.cloudbreak.service.stack.StackService;
+import com.sequenceiq.cloudbreak.util.NodesUnreachableException;
+import com.sequenceiq.cloudbreak.util.StackUtil;
+
+@Component
+public class SaltCheckerConclusionStep extends ConclusionStep {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(SaltCheckerConclusionStep.class);
+
+    @Inject
+    private NodeStatusService nodeStatusService;
+
+    @Inject
+    private StackService stackService;
+
+    @Inject
+    private StackUtil stackUtil;
+
+    @Override
+    public Conclusion check(Long resourceId) {
+        RPCResponse<SaltHealthReport> saltPingResponse;
+        try {
+            saltPingResponse = nodeStatusService.saltPing(resourceId);
+        } catch (CloudbreakServiceException e) {
+            LOGGER.debug("Salt health report failed, fallback and check unreachable nodes, error: {}", e.getMessage());
+            return checkUnreachableNodes(resourceId);
+        }
+        SaltHealthReport report = saltPingResponse.getResult();
+        if (report == null) {
+            LOGGER.debug("Salt health report was null, fallback and check unreachable nodes.");
+            return checkUnreachableNodes(resourceId);
+        }
+
+        List<String> failedServicesOnMaster = collectFailedServicesOnMaster(report);
+        if (!failedServicesOnMaster.isEmpty()) {
+            return createFailedConclusionForMaster(failedServicesOnMaster);
+        } else {
+            Map<String, String> unreachableMinions = collectUnreachableMinions(report);
+            if (unreachableMinions.isEmpty()) {
+                return succeeded();
+            } else {
+                return createFailedConclusionForMinions(unreachableMinions);
+            }
+        }
+    }
+
+    private Conclusion checkUnreachableNodes(Long resourceId) {
+        Stack stack = stackService.getByIdWithListsInTransaction(resourceId);
+        Set<String> allNodes = stackUtil.collectNodes(stack).stream().map(Node::getHostname).collect(Collectors.toSet());
+        try {
+            stackUtil.collectAndCheckReachableNodes(stack, allNodes);
+        } catch (NodesUnreachableException e) {
+            Set<String> unreachableNodes = e.getUnreachableNodes();
+            String conclusion = String.format("Unreachable nodes: %s. Please check the instances on your cloud provider for further details.",
```

Comment:
I assume this message might get to the customer;
Unreachable nodes can be ambiguous, I would try to clarify the message here, where does the connection problematic, from where we can't reach nodes.

- Is it CDP can't reach the cluster nodes?  (here it might be useful to have another level of digdown, what's not available exactly.)
- Is it a connection problem inside the Cluster and nodes can't talk to each other.